arbos/l2pricing/model.go

```diff
@@ -26,100 +28,203 @@ const InitialPricingInertia = 102
+	gasModel, err := ps.GasModelToUse()
+	if err != nil {
+		return err
+	}
+	switch gasModel {
+	case GasModelLegacy:
+		return ps.updateLegacyBacklog(true, usedGas)
+	case GasModelSingleGasConstraints:
+		return ps.updateSingleGasConstraintsBacklogs(true, usedGas)
+	case GasModelMultiGasConstraints:
+		return ps.updateMultiGasConstraintsBacklogs(true, usedGas, usedMultiGas)
+	default:
+		return fmt.Errorf("can not determine gas model")
 	}
-	return false, nil
 }
 
-func (ps *L2PricingState) AddToGasPool(gas int64) error {
-	shouldUseGasConstraints, err := ps.ShouldUseGasConstraints()
+// ShrinkBacklog reduces the backlog for the active pricing model.
+func (ps *L2PricingState) ShrinkBacklog(usedGas uint64, usedMultiGas multigas.MultiGas) error {
+	gasModel, err := ps.GasModelToUse()
 	if err != nil {
 		return err
 	}
-	if shouldUseGasConstraints {
-		return ps.addToGasPoolWithGasConstraints(gas)
+	switch gasModel {
+	case GasModelLegacy:
+		return ps.updateLegacyBacklog(false, usedGas)
+	case GasModelSingleGasConstraints:
+		return ps.updateSingleGasConstraintsBacklogs(false, usedGas)
+	case GasModelMultiGasConstraints:
+		return ps.updateMultiGasConstraintsBacklogs(false, usedGas, usedMultiGas)
+	default:
+		return fmt.Errorf("can not determine gas model")
 	}
-	return ps.addToGasPoolLegacy(gas)
 }
 
-func (ps *L2PricingState) addToGasPoolLegacy(gas int64) error {
+func (ps *L2PricingState) updateLegacyBacklog(growBacklog bool, usedGas uint64) error {
 	backlog, err := ps.GasBacklog()
 	if err != nil {
 		return err
 	}
-	backlog = applyGasDelta(backlog, gas)
+	backlog = applyGasDelta(backlog, growBacklog, usedGas)
 	return ps.SetGasBacklog(backlog)
 }
 
-func (ps *L2PricingState) addToGasPoolWithGasConstraints(gas int64) error {
+func (ps *L2PricingState) updateSingleGasConstraintsBacklogs(growBacklog bool, usedGas uint64) error {
 	constraintsLength, err := ps.gasConstraints.Length()
 	if err != nil {
-		return fmt.Errorf("failed to get number of constraints: %w", err)
+		return err
 	}
 	for i := range constraintsLength {
 		constraint := ps.OpenGasConstraintAt(i)
 		backlog, err := constraint.Backlog()
 		if err != nil {
-			return fmt.Errorf("failed to get backlog of constraint %v: %w", i, err)
+			return err
 		}
-		err = constraint.SetBacklog(applyGasDelta(backlog, gas))
+		err = constraint.SetBacklog(applyGasDelta(backlog, growBacklog, usedGas))
 		if err != nil {
-			return fmt.Errorf("failed to set backlog of constraint %v: %w", i, err)
+			return err
 		}
 	}
 	return nil
 }
 
-func (ps *L2PricingState) GasPoolUpdateCost() uint64 {
-	result := storage.StorageReadCost + storage.StorageWriteCost
+func (ps *L2PricingState) updateMultiGasConstraintsBacklogs(growBacklog bool, usedGas uint64, usedMultiGas multigas.MultiGas) error {
+	if usedMultiGas.SingleGas() != usedGas {
+		log.Warn("usedGas does not match sum of usedMultiGas", "usedGas", usedGas, "usedMultiGas", usedMultiGas.SingleGas())
+	}
```

Comment:
It is so easy to just ignore this warning. I'd rather remove this and think of a better way of testing that these numbers won't diverge.

Reply:
Also, by the time we see this warning in production, it will be too late to fix it because we will have to support this divergence forever.

Reply:
Removed for now, probably there is way to have a debug-only assertion